openmls/src/treesync/treesync_node.rs

```diff
@@ -145,21 +147,46 @@ impl TreeSyncParentNode {
         &mut self.node
     }
 
-    /// Compute the tree hash for this node, thus populating the `tree_hash`
-    /// field.
+    /// Compute the tree hash for this node. Leaf nodes from the exclusion list
+    /// are filtered out.
     pub(in crate::treesync) fn compute_tree_hash(
         &self,
         backend: &impl OpenMlsCryptoProvider,
         ciphersuite: Ciphersuite,
         left_hash: Vec<u8>,
         right_hash: Vec<u8>,
+        exclusion_list: &HashSet<&LeafNodeIndex>,
     ) -> Result<Vec<u8>, LibraryError> {
-        let hash_input = TreeHashInput::new_parent(
-            self.node.as_ref(),
-            VLByteSlice(&left_hash),
-            VLByteSlice(&right_hash),
-        );
-        let hash = hash_input.hash(backend, ciphersuite)?;
+        let hash = if exclusion_list.is_empty() {
+            // If the exclusion list is empty, we can just use the parent node
+            TreeHashInput::new_parent(
+                self.node.as_ref(),
+                VLByteSlice(&left_hash),
+                VLByteSlice(&right_hash),
+            )
+            .hash(backend, ciphersuite)?
+        } else if let Some(parent_node) = self.node.as_ref() {
+            // If the exclusion list is not empty, we need to create a new
+            // parent node without the excluded indices in the unmerged leaves.
+            let mut new_node = parent_node.clone();
+            let unmerged_leaves = new_node
+                .unmerged_leaves()
+                .iter()
+                .filter(|leaf| !exclusion_list.contains(leaf))
+                .cloned()
+                .collect();
+            new_node.set_unmerged_leaves(unmerged_leaves);
+            TreeHashInput::new_parent(
+                Some(&new_node),
+                VLByteSlice(&left_hash),
+                VLByteSlice(&right_hash),
+            )
+            .hash(backend, ciphersuite)?
+        } else {
+            // If the node is blank
+            TreeHashInput::new_parent(None, VLByteSlice(&left_hash), VLByteSlice(&right_hash))
+                .hash(backend, ciphersuite)?
+        };
```

Comment:
As discussed OOB there are some minor refactoring opportunities here around clone, but the impact will probably be minimal. (Although this function will be executed a number of times during the loop in parent hash verification.)

Reply:
I looked into it and the change would require going through (and refactoring) many more layers until we can isolate unmerged leaves. The benefit would be that we only clone unmerged leaves and not the HPKE public key of the node.
I decided that the added code complexity was not worth the benefit. We can revisit this kind of optimisation in the context of performance improvements, right now this is in line with other performance levels.